File: gastromanager-backend/src/main/java/com/kaiho/gastromanager/application/ingredient/mapper/IngredientMapper.java
package com.kaiho.gastromanager.application.ingredient.mapper;

import com.kaiho.gastromanager.application.ingredient.dto.request.IngredientRequestDto;
import com.kaiho.gastromanager.application.ingredient.dto.request.IngredientUpdateRequestDto;
import com.kaiho.gastromanager.application.ingredient.dto.response.IngredientDetailResponseDto;
import com.kaiho.gastromanager.application.ingredient.dto.response.IngredientSummaryResponseDto;
import com.kaiho.gastromanager.domain.ingredient.model.Ingredient;
import com.kaiho.gastromanager.domain.ingredient.model.Unit;
import org.springframework.stereotype.Component;

@Component
public class IngredientMapper {

    public IngredientSummaryResponseDto toResponseSummary(Ingredient ingredient) {
        if (ingredient == null) {
            return null;
        }

        return IngredientSummaryResponseDto.builder()
                                           .uuid(ingredient.getUuid())
                                           .name(ingredient.getName())
                                           .supplier(ingredient.getSupplier())
                                           .minimumStockQuantity(ingredient.getMinimumStockQuantity())
                                           .unit(ingredient.getUnit().getSymbol())
                                           .pricePerUnit(ingredient.getPricePerUnit())
                                           .availableStock(ingredient.getAvailableStock())
                                           .updatedDate(ingredient.getUpdatedDate())
                                           .updatedBy(ingredient.getUpdatedBy())
                                           .createdBy(ingredient.getCreatedBy())
                                           .createdDate(ingredient.getCreatedDate())
                                           .build();
    }

    public IngredientDetailResponseDto toResponseDetail(Ingredient ingredient) {
        if (ingredient == null) {
            return null;
        }

        return IngredientDetailResponseDto.builder()
                                          .uuid(ingredient.getUuid())
                                          .name(ingredient.getName())
                                          .supplier(ingredient.getSupplier())
                                          .minimumStockQuantity(ingredient.getMinimumStockQuantity())
                                          .unit(ingredient.getUnit().getSymbol())
                                          .pricePerUnit(ingredient.getPricePerUnit())
                                          .availableStock(ingredient.getAvailableStock())
                                          .updatedDate(ingredient.getUpdatedDate())
                                          .updatedBy(ingredient.getUpdatedBy())
                                          .createdBy(ingredient.getCreatedBy())
                                          .createdDate(ingredient.getCreatedDate())
                                          .build();
    }

    public Ingredient toDomain(IngredientRequestDto ingredientRequestDto) {
        if (ingredientRequestDto == null) {
            return null;
        }
        return Ingredient.builder()
                         .name(ingredientRequestDto.name())
                         .availableStock(ingredientRequestDto.availableStock())
                         .unit(Unit.valueOf(ingredientRequestDto.unit()))
                         .pricePerUnit(ingredientRequestDto.pricePerUnit())
                         .minimumStockQuantity(ingredientRequestDto.minimumStockQuantity())
                         .supplier(ingredientRequestDto.supplier())
                         .build();
    }

    public Ingredient toDomain(IngredientUpdateRequestDto ingredientRequestDto) {
        if (ingredientRequestDto == null) {
            return null;
        }
        return Ingredient.builder()
                         .name(ingredientRequestDto.name())
                         .unit(Unit.valueOf(ingredientRequestDto.unit()))
                         .pricePerUnit(ingredientRequestDto.pricePerUnit())
                         .minimumStockQuantity(ingredientRequestDto.minimumStockQuantity())
                         .supplier(ingredientRequestDto.supplier())
                         .build();
    }

}

File: gastromanager-frontend/src/app/features/ingredients/application/mappers/ingredient.mapper.ts
import {IngredientResponseDto} from "../../domain/models/ingredient-response-dto.interface";
import {Ingredient} from "../../domain/models/ingredient.interface";
import {IngredientRequestDto} from "../../domain/models/ingredient-request-dto.interface";
import {IngredientDetailResponseDto} from "../../domain/models/ingredient-detail-response-dto.interface";


export function mapToIngredientSummary(dto: IngredientResponseDto): Ingredient {
  return {
    uuid: dto.uuid,
    name: dto.name,
    pricePerUnit: dto.pricePerUnit,
    supplier: dto.supplier,
    unit: dto.unit,
    availableStock: dto.availableStock,
    minimumStockQuantity: dto.minimumStockQuantity,
    createdBy: dto.createdBy,
    createdDate: dto.createdDate,
    updatedBy: dto.updatedBy,
    updatedDate: dto.updatedDate,
  };
}

export function mapToIngredientDetail(dto: IngredientDetailResponseDto): Ingredient {
  return {
    uuid: dto.uuid,
    name: dto.name,
    pricePerUnit: dto.pricePerUnit,
    supplier: dto.supplier,
    unit: dto.unit,
    availableStock: dto.availableStock,
    minimumStockQuantity: dto.minimumStockQuantity,
    createdBy: dto.createdBy,
    createdDate: new Date(dto.createdDate),
    updatedBy: dto.updatedBy,
    updatedDate: new Date(dto.updatedDate)
  };
}

export function mapToIngredientRequestDto(ingredient: Ingredient): IngredientRequestDto {
  return {
    name: ingredient.name,
    pricePerUnit: ingredient.pricePerUnit,
    availableStock: ingredient.availableStock,
    unit: ingredient.unit,
    supplier: ingredient.supplier,
    minimumStockQuantity: ingredient.minimumStockQuantity
  };
}

export function mapToEditIngredientRequestDto(ingredient: Ingredient): IngredientRequestDto {
  return {
    name: ingredient.name,
    pricePerUnit: ingredient.pricePerUnit,
    unit: ingredient.unit,
    supplier: ingredient.supplier,
    minimumStockQuantity: ingredient.minimumStockQuantity
  };
}

